File: Data Structures & Algorithms/binary-tree-diameter/submission-1.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func diameter(root *TreeNode, d *int) int{
	if root == nil{
		return 0
	}

	left := diameter(root.Left, d)
	right := diameter(root.Right, d)

	*d = max(*d, left+right) 

	return 1+ max(left, right)
}

func diameterOfBinaryTree(root *TreeNode) int {

	var d = 0

	diameter(root, &d)
	return d
    
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def max_depth(self, root):
        if root == None:
            return 0

        height = 1 + max(self.max_depth(root.left), self.max_depth(root.right))
        return height

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root == None:
            return True

        left_depth = self.max_depth(root.left) 
        right_depth = self.max_depth(root.right)

        if abs(left_depth - right_depth) > 1:
            return False

        return self.isBalanced(root.left) and self.isBalanced(root.right)




